### ISSUE TITLE
loadLibrary method on deferred imports should return `Future<void>` instead of `Future<dynamic>`

### ISSUE THREAD

**Issue by jonahwilliams** [dart-lang/sdk#38379]
Currently the future returned by calling `loadLibrary` on a deferred import returns a `Future<dynamic>`. Since I've not used deferred imports before, I assumed that this value was what I would use to interact with the deferred library. I was surprised to discover that the library should be used through the original import prefix instead. I think it would be less confusing initially if it returned `void`, as the narrows down the possible options significantly.

**Comment by devoncarew** [comment]
@jonahwilliams - this is `loadLibrary()` from `dart:mirrors`?

**Comment by jonahwilliams** [comment]
No, the dart2js deferred import syntax:

https://news.dartlang.org/2014/08/dart-16-adds-support-for-deferred.html

**Comment by sigmundch** [comment]
At this time `loadLibrary` is injected by the CFE, so I believe this should only require a fix there.

I like your suggestion, my only concern is whether this would constitute a breaking change (I believe technically it is, but that in practice it might be OK). If so we need to go through the proper process to make sure the effect on existing users is none or minimal.

**Comment by jonahwilliams** [comment]
Well, technically everything is breaking. I'd be happy to drive the breaking change process forward here. Though its minor, I believe it would constitute a reasonable improvement in the API.

Next week I can take a look at how this is implemented to gather an idea on how breaking it would be.

**Comment by eernstg** [comment]
I agree that `Future<void>` describes the situation better than `Future<dynamic>` does. The signature of `loadLibrary` was not mentioned in the language specification previously, and dart-lang/language#567 plus dart-lang/language#568 brings up exactly this issue.

**Comment by sigmundch** [comment]
> Next week I can take a look at how this is implemented to gather an idea on how breaking it would be.

Note, we simply create a top-level hidden method called `__loadLibrary_prefix` for each import with a deferred prefix. The declaration has the `Future<dynamic>` return type declared:

https://github.com/dart-lang/sdk/blob/b0fa0aebb7064b3af60a7504df832fe3bf4c9a44/pkg/front_end/lib/src/fasta/kernel/load_library_builder.dart#L58

So if we move forward with this change, it would be a pretty small edit to fix this.

**Comment by sigmundch** [comment]
In terms of breaking change risk - I should note that when I first implemented this in the CFE I had not included a type and it had defaulted to `dynamic` instead of `Future<dynamic>`, we changed it then because users had noticed the change in static errors: https://github.com/dart-lang/sdk/issues/32604

We can use intl again to validate that this change will not cause new issues.

**Comment by jonahwilliams** [comment]
I did not end up getting to this last week. I'll plan on taking a look with package:intl though
